### PR TITLE
[WFLY-6143] Provide ability to add a server-side configured interceptor for the applications without changing the deployments

### PR DESCRIPTION
**Pull request by zhantemirov**
A test case that covers global EJB server side interceptors configuration feature.

First test case tests the following scenarios:
* pack an interceptor module and place it to $JBOSS_HOME/modules directory
* check adding a value to server-interceptors node in ejb3 subsystem
* reload the server so that the interceptor takes effect

The second and last test case should use CountDownLatch class for the server-side interceptor methods firing check.

Jira issue: https://issues.jboss.org/browse/WFLY-6143
Analysis doc: https://github.com/wildfly/wildfly-proposals/pull/180/

### CODE REVIEW

**Comment by bstansberry** [comment]
@zhantemirov Please coordinate with @tadamski to get this included in his PR #12268.

**Comment by zhantemirov** [comment]
@bstansberry sure, we have already started discussion on it, thanks.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.InvocationContext;
+import org.jboss.logging.Logger;
+
+/**
+ * Server side sample interceptor
+ */
+public class ServerInterceptor {
+
+    public ServerInterceptor() {
```
Comment (tadamski) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java]: We need to make sure that the interceptor was executed. static CountDownLatch possibly?

**Review by tadamski — COMMENTED** [review]
I would also add the @AroundTimeout test case

**Comment by zhantemirov** [comment]
@mjurc @tadamski please feel free to review and provide your opinions regarding the methods and decisions I made in this test case. Thanks.

**Review by tadamski — COMMENTED** [review]
@zhantemirov please rabase the test on top of tadamski:WFLY-6143 branch.
Currently, the test fails after the rebase because of configuration issues.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    private Logger log = Logger.getLogger(ServerInterceptor.class.getName());
+
+    @AroundInvoke
+    public Object aroundInvoke(final InvocationContext invocationContext) throws Exception {
+        log.debug("serverInterceptor aroundInvoke ");
```
Comment (tadamski) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java]: We need to check that the interceptor was executed.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerSideDefinedInterceptorTestCase.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+/**
+ * A test case verifying an ability of adding a server-side configured interceptor without changes deployments.
+ * See https://issues.jboss.org/browse/WFLY-6143 for more details.
+ *
+ * @author <a href="mailto:szhantem@redhat.com">Sultan Zhantemirov</a> (c) 2019 Red Hat, inc.
+ */
+@RunWith(Arquillian.class)
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class ServerSideDefinedInterceptorTestCase {
```
Comment (tadamski) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerSideDefinedInterceptorTestCase.java]: That's cosmetic but I think ServerInterceptorsTestCase would be more consistent with test naming used in testsuite

Comment (zhantemirov) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerSideDefinedInterceptorTestCase.java]: Thanks for the note, renamed.

**Comment by zhantemirov** [comment]
@tadamski PR has been rebased. 
Going to implement the second test case using CountDownLatch.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java (hunk 1 above)
Comment (zhantemirov) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java]: Added, could you please review it? Thanks.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.InvocationContext;
+
+/**
+ * Server side sample interceptor
+ * @author <a href="mailto:szhantem@redhat.com">Sultan Zhantemirov</a> (c) 2019 Red Hat, inc.
+ */
+public class ServerInterceptor {
+
+    public final static CountDownLatch latch = new CountDownLatch(1);
```
Comment (tadamski) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java]: public static final (checkstyle)

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptorsExecutionTestCase.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                    return bean.getSimpleName();
+                }
+            });
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        try {
+            // waiting for the interceptor's method execution
+            ServerInterceptor.latch.await();
```
Comment (tadamski) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptorsExecutionTestCase.java]: Do we need to execute it asynchronously? Furthermore, we can't wait indefinitely on the latch.
I would  execute it synchronously and later just run
ServerInterceptor.latch.getCount()==0.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/module.xml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module xmlns="urn:jboss:module:1.5" name="interceptor-module">
+    <resources>
+        <resource-root path="server-side-interceptor.jar"/>
+    </resources>
+    <dependencies>
+        <module name="javax.interceptor.api"/>
+    </dependencies>
+</module>
```
Comment (tadamski) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/module.xml]: hmm I guess we need one more endline

**Review by tadamski — CHANGES_REQUESTED** [review]

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java [hunk 7]
```diff
@@ -23,15 +23,19 @@
...
 
 /**
  * Server side sample interceptor
  * @author <a href="mailto:szhantem@redhat.com">Sultan Zhantemirov</a> (c) 2019 Red Hat, inc.
  */
+@Interceptor
 public class ServerInterceptor {
 
     public final static CountDownLatch latch = new CountDownLatch(1);
+    public final static CountDownLatch timeoutLatch = new CountDownLatch(1);
```
Comment (tadamski) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java]: static final (checkstyle)

**Review by tadamski — CHANGES_REQUESTED** [review]

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java (hunk 7 above)
Comment (zhantemirov) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java]: Fixed, thanks.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/module.xml (hunk 6 above)
Comment (zhantemirov) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/module.xml]: Fixed, thanks.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java (hunk 4 above)
Comment (zhantemirov) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java]: Fixed, thanks.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java (hunk 2 above)
Comment (zhantemirov) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptor.java]: Added checking with CountDownLatch

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptorsExecutionTestCase.java (hunk 5 above)
Comment (zhantemirov) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/interceptor/serverside/ServerInterceptorsExecutionTestCase.java]: Thank you for the suggestion, updated.

**Comment by jamezp** [comment]
Closing based on JIRA comments.

**Pull request closed without merging**